service/history/consistency_checker.go

```diff
@@ -0,0 +1,313 @@
+		ctx,
+		namespace.ID(workflowKey.NamespaceID),
+		commonpb.WorkflowExecution{
+			WorkflowId: workflowKey.WorkflowID,
+			RunId:      workflowKey.RunID,
+		},
+		workflow.CallerTypeAPI,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	mutableState, err := wfContext.LoadWorkflowExecution(ctx)
+	switch err.(type) {
+	case nil:
+		if consistencyPredicate(mutableState) {
+			return newWorkflowContext(wfContext, release, mutableState), nil
+		}
+
+		wfContext.Clear()
+		mutableState, err := wfContext.LoadWorkflowExecution(ctx)
+		if err != nil {
+			release(err)
+			return nil, err
+		}
+		return newWorkflowContext(wfContext, release, mutableState), nil
+	case *serviceerror.NotFound:
+		release(err)
+		if err := assertShardOwnership(
+			ctx,
+			c.shardContext,
+		); err != nil {
+			return nil, err
+		}
+		return nil, err
+	default:
+		release(err)
+		return nil, err
+	}
+}
+
+func (c *WorkflowConsistencyCheckerImpl) getCurrentWorkflowContext(
+	ctx context.Context,
+	consistencyPredicate MutableStateConsistencyPredicate,
+	namespaceID string,
+	workflowID string,
+) (workflowContext, error) {
+	for attempt := 1; attempt <= conditionalRetryCount; attempt++ {
+		runID, err := c.getCurrentRunID(
+			ctx,
+			namespaceID,
+			workflowID,
+		)
+		if err != nil {
+			return nil, err
+		}
+		wfContext, err := c.getWorkflowContextValidatedByCheckFns(
+			ctx,
+			consistencyPredicate,
+			definition.NewWorkflowKey(namespaceID, workflowID, runID),
+		)
+		if err != nil {
+			return nil, err
+		}
+		if wfContext.getMutableState().IsWorkflowExecutionRunning() {
+			return wfContext, nil
+		}
+
+		currentRunID, err := c.getCurrentRunID(
+			ctx,
+			namespaceID,
+			workflowID,
+		)
+		if err != nil {
+			wfContext.getReleaseFn()(err)
+			return nil, err
+		}
+		if currentRunID == wfContext.getRunID() {
+			return wfContext, nil
+		}
+		wfContext.getReleaseFn()(nil)
+	}
+	return nil, serviceerror.NewUnavailable("unable to locate current workflow execution")
+}
+
+func (c *WorkflowConsistencyCheckerImpl) getCurrentRunID(
+	ctx context.Context,
+	namespaceID string,
+	workflowID string,
+) (string, error) {
+	resp, err := c.shardContext.GetCurrentExecution(
+		ctx,
+		&persistence.GetCurrentExecutionRequest{
+			ShardID:     c.shardContext.GetShardID(),
+			NamespaceID: namespaceID,
+			WorkflowID:  workflowID,
+		},
+	)
+	switch err.(type) {
+	case nil:
+		return resp.RunID, nil
+	case *serviceerror.NotFound:
+		if err := assertShardOwnership(
+			ctx,
+			c.shardContext,
+		); err != nil {
+			return "", err
+		}
+		return "", err
+	default:
+		return "", err
+	}
+}
+
+func assertShardOwnership(
+	ctx context.Context,
+	shardContext shard.Context,
+) error {
+	consistencyChecked := ctx.Value(consistencyCtxKey).(consistencyContextValue)
+	if !consistencyChecked.checked {
+		consistencyChecked.checked = true
+		return shardContext.AssertOwnership(ctx)
+	}
+	return nil
+}
+
+func BypassMutableStateConsistencyPredicate(
```

Comment:
I see this is used in most of places, including RespondActivityTaskCompleted where the action of updateWorkflowActionFunc still rely on scheduleID >= mutableState.GetNextEventID() to detect stale cache do we still need those? I think the intention is to rely on the vclock to detect the staleness in those cases?

Reply:
>I see this is used in most of places, including RespondActivityTaskCompleted where the action of updateWorkflowActionFunc still rely on scheduleID >= mutableState.GetNextEventID() to detect stale cache do we still need those? I think the intention is to rely on the vclock to detect the staleness in those cases?

currently only respond workflow task completed will use a customized predicate, i will probably update other place after this huge PR

>I think the intention is to rely on the vclock to detect the staleness in those cases?
yes, but there will be a transition period: forward / backward compatibility